trio/_core/_run.py

```diff
@@ -295,57 +293,47 @@ def started(self, value=None):
         self._old_nursery._check_nursery_closed()
 
 
-@asynccontextmanager
-@async_generator
-@enable_ki_protection
-async def open_nursery():
-    """Returns an async context manager which creates a new nursery.
+class NurseryManager:
+    def __init__(self, wrapper_func_name):
+        self._wrapper_func_name = wrapper_func_name
 
-    This context manager's ``__aenter__`` method executes synchronously. Its
-    ``__aexit__`` method blocks until all child tasks have exited.
+    @enable_ki_protection
+    async def __aenter__(self):
+        assert currently_ki_protected()
+        self._scope_manager = open_cancel_scope()
+        scope = self._scope_manager.__enter__()
+        self._nursery = Nursery(current_task(), scope)
+        return self._nursery
 
-    """
-    assert currently_ki_protected()
-    with open_cancel_scope() as scope:
-        nursery = Nursery(current_task(), scope)
-        nested_child_exc = None
-        try:
-            await yield_(nursery)
-        except BaseException as exc:
-            nested_child_exc = exc
+    @enable_ki_protection
+    async def __aexit__(self, etype, exc, tb):
         assert currently_ki_protected()
-        await nursery._nested_child_finished(nested_child_exc)
-
-
-# I *think* this is equivalent to the above, and it gives *much* nicer
-# exception tracebacks... but I'm a little nervous about it because it's much
-# trickier code :-(
-#
-# class NurseryManager:
-#     @enable_ki_protection
-#     async def __aenter__(self):
-#         self._scope_manager = open_cancel_scope()
-#         scope = self._scope_manager.__enter__()
-#         self._parent_nursery = Nursery(current_task(), scope)
-#         return self._parent_nursery
-#
-#     @enable_ki_protection
-#     async def __aexit__(self, etype, exc, tb):
-#         try:
-#             await self._parent_nursery._clean_up(exc)
-#         except BaseException as new_exc:
-#             if not self._scope_manager.__exit__(
-#                     type(new_exc), new_exc, new_exc.__traceback__):
-#                 if exc is new_exc:
-#                     return False
-#                 else:
-#                     raise
-#         else:
-#             self._scope_manager.__exit__(None, None, None)
-#             return True
-#
-# def open_nursery():
-#     return NurseryManager()
+        nested_child_exc = exc if isinstance(exc, BaseException) else None
```

Comment:
This line is equivalent to `nested_child_exc = exc` :-). The context manager protocol says that `exc` can be either an exception object, or `None`. And all exception objects are instances of `BaseException`, by definition.